x-pack/plugin/security/src/main/java/org/elasticsearch/xpack/security/authz/store/NativePrivilegeStore.java

```diff
@@ -83,47 +100,120 @@
     }
 
     public void getPrivileges(Collection<String> applications, Collection<String> names,
                               ActionListener<Collection<ApplicationPrivilegeDescriptor>> listener) {
+
+        // TODO: We should have a way to express true Zero applications
+        final Set<String> applicationNamesCacheKey = (isEmpty(applications) || applications.contains("*")) ?
+            Collections.singleton("*") : Set.copyOf(applications);
+
+        // Always fetch for the concrete application names even when the passed in application names do not
+        // contain any wildcard. This serves as a negative lookup.
+        Set<String> concreteApplicationNames = applicationNamesCache.get(applicationNamesCacheKey);
+
+        if (concreteApplicationNames != null && concreteApplicationNames.size() == 0) {
+            logger.debug("returning empty application privileges as application names result in empty list");
+            listener.onResponse(Collections.emptySet());
+
+        } else {
+            final Tuple<Set<String>, Map<String, Set<ApplicationPrivilegeDescriptor>>> cacheStatus;
+            if (concreteApplicationNames == null) {
+                cacheStatus = cacheStatusForApplicationNames(applicationNamesCacheKey);
+            } else {
+                cacheStatus = cacheStatusForApplicationNames(concreteApplicationNames);
+            }
+
+            if (cacheStatus.v1().isEmpty()) {
+                logger.debug("All application privileges found in cache");
+                final Set<ApplicationPrivilegeDescriptor> cachedDescriptors =
+                    cacheStatus.v2().values().stream().flatMap(Collection::stream).collect(Collectors.toUnmodifiableSet());
+                listener.onResponse(filterDescriptorsForNames(cachedDescriptors, names));
+            } else {
+                final long invalidationCounter = numInvalidation.get();
+                // Always fetch all privileges of an application for caching purpose
+                logger.debug("Fetching application privilege documents for: {}", cacheStatus.v1());
+                innerGetPrivileges(cacheStatus.v1(), ActionListener.wrap(fetchedDescriptors -> {
+                    final Map<String, Set<ApplicationPrivilegeDescriptor>> mapOfFetchedDescriptors = fetchedDescriptors.stream()
+                        .collect(Collectors.groupingBy(ApplicationPrivilegeDescriptor::getApplication, Collectors.toUnmodifiableSet()));
+                    final Set<String> allApplicationNames =
+                        Stream.concat(cacheStatus.v2().keySet().stream(), mapOfFetchedDescriptors.keySet().stream())
+                            .collect(Collectors.toUnmodifiableSet());
+                    // Avoid caching potential stale results.
+                    // TODO: It is still possible that cache gets invalidated immediately after the if check
+                    if (invalidationCounter == numInvalidation.get()) {
```

Comment:
This pattern is used in `CompositeRolesStore`. However it is still possible that the cache gets invalidated immediately after the `if` check. The window for it to happen is much small, but still exists in theory.

The other pattern is to use `ListenableFuture`. This would solves the stale entry problem here because the `future` itself is removed from the cached. So adding items to the removed `future` has no impact to the cache. However, this pattern could potentially have a deadlock issue? If the thread computing for future crashes, will all the other thread waiting for it get stuck? If this is true, I'd prefer to use the first pattern since a stale entry (with very low chance) is less harmful than deadlocking.

Reply:
I think a ReadWriteLock would solve the problem. Treat the invalidator as the writer and the cache population as the reader. The invalidator would need exclusive access, but we could support multiple populators.

I'll need to think about it again when it's not midnight, but I think it's reasonable (if used alongside the `invalidationCounter` so the lock window is small).